include/fluent-bit/flb_config.h

```diff
@@ -320,6 +322,7 @@ enum conf_type {
 #define FLB_CONF_STR_PLUGINS_FILE "Plugins_File"
 #define FLB_CONF_STR_STREAMS_FILE "Streams_File"
 #define FLB_CONF_STR_CONV_NAN     "json.convert_nan_to_null"
+#define FLB_CONF_ROUTES_MASK_ELEMENTS "Routes_Mask_Elements"
```

Comment:
The side effect of setting routing mask size is that you'd end up with an upper limit of outputs (256). But maybe instead of exposing this size, what if something like this is done:

- Define a config option for upper limit of number of outputs. Let's call it X.
- Calculate the size of routes array from X: ceil(X/64)

For internal implementation I can do what you're suggesting with wrapping routes array along with size in a struct.
wdyt?

Reply:
When I say routing mask size I mean that in a loose way, it's not about bits but rather about T size elements required to fit N bits (in this case "how many unsigned 64 bit integers we need in the array").

I don't think we need to define a configuration option, the user doesn't need to know because that's a pure runtime issue that directly correlates to the amount of outputs defined in the configuration file.

As far as wrapping the array and size with a struct I agree, that's what I asked for, that way it's self contained and the functions that read/write routing don't need access to any more context than the route instance and bit position (if applicable).

I'll include a very basic example just to be sure nothing is getting lost in translation : 

```
struct flb_route {
    size_t size;
    uint64_t *data;
};
```

An alternative approach where the data member is an open array at the end of the structure and the actual route is allocated taking the entire size in account (static + variable) looks somewhat neater but I think it'd be wiser to lean on the conservative side in this case if we want this change to land and to minimize the risk of causing issues.

Reply:
Sorry, but I'm not seeing how this will solve the problem of having a hard coded limit of 256 outputs. I want to be able to configure many 100s or 1000s of outputs depending on my need and resources, and unless I'm missing something not providing a configuration option doesn't solve that case. Please correct me if I'm wrong here.

Reply:
The idea is that if you configure 1000 outputs fluent-bit will detect that in the initialization phase and will set the internal routing mask size appropriately. 

Why should the user need to know about the routing mask size when fluent-bit clearly know exactly how large it needs to be.

If you need a code reference let me know or feel free to message me in slack if you want a more in depth explanation or to make a short call (well, tomorrow since it's pretty late for me now).

Reply:
Ah, I see, this is better. I'll give it a shot and will reach out in case I'm stumped. Thanks for all the pointers and suggestions!

Reply:
Super cool!